### DISCUSSION TITLE
Define the app's InitialRoute depending on a provider

### DISCUSSION THREAD

**Discussion by MichaelLucca** [rrousselGit/riverpod#3414] [Q&A]
**Description**
Currently, I'm using a single provider in Flutter to fetch data from multiple API calls within my application. However, I'm encountering difficulty in managing the progressive display of Data and Errors in my UI.

Currently, I'm using a provider in Flutter to fetch some data with FlutterSecureStorage. Based on the result of the operation, I want to send the user on a different route.
However, in release mode the state is correctly updated, but the initialRoute doesn't change.

The code I'm currently dealing with:
![Screenshot 2024-03-11 alle 15 50 13](https://github.com/rrousselGit/riverpod/assets/89570547/b0dd2ec5-9548-4d65-a2b5-d2e5d9829fd5)

![Screenshot 2024-03-11 alle 15 50 41](https://github.com/rrousselGit/riverpod/assets/89570547/5fd1c2c2-c1ce-4604-af35-ce7c610391e6)

**Comment by Colman** [comment]
If I'm not mistaken, the `initialRoute` parameter is only used on the first build of the `MaterialApp` widget. So if the `initialRoute` is `routeA` on the first build, then after that the `when` call starts returning `routeB`, the app won't redirect to `routeB`.

If you need to do some async computation to determine what your initial route will be, what do you plan on displaying during this computation? You could either show a custom loading screen, in which case that loading screen would be your initial route, or you could just show the app splash screen. If you make the `main` function in `main.dart` asynchronous, it will show the app splash screen until you call `runApp(...)`. So maybe you could do your computation there and pass the initial route to the widget in `runApp`.
